clar2wasm/src/duck_type.rs

```diff
@@ -0,0 +1,585 @@
+        og_ty: &TypeSignature,
+        target_ty: &TypeSignature,
+        locals: &[LocalId],
+    ) -> Result<(), GeneratorError> {
+        match (og_ty, target_ty) {
+            (TypeSignature::NoType, _) | (_, TypeSignature::NoType) => {
+                // drop the useless original value from the stack
+                drop_value(builder, og_ty);
+                // set locals to zero values (needed for ducktying of list elements)
+                add_placeholder_for_clarity_type(builder, target_ty);
+                for &l in locals.iter().rev() {
+                    builder.local_set(l);
+                }
+            }
+            (TypeSignature::BoolType, TypeSignature::BoolType)
+            | (TypeSignature::IntType, TypeSignature::IntType)
+            | (TypeSignature::UIntType, TypeSignature::UIntType)
+            | (TypeSignature::PrincipalType, TypeSignature::PrincipalType)
+            | (
+                TypeSignature::SequenceType(SequenceSubtype::BufferType(_)),
+                TypeSignature::SequenceType(SequenceSubtype::BufferType(_)),
+            )
+            | (
+                TypeSignature::SequenceType(SequenceSubtype::StringType(_)),
+                TypeSignature::SequenceType(SequenceSubtype::StringType(_)),
+            )
+            | (TypeSignature::CallableType(_), TypeSignature::CallableType(_))
+            | (TypeSignature::TraitReferenceType(_), TypeSignature::TraitReferenceType(_)) => {
+                for &l in locals.iter().rev() {
+                    builder.local_set(l);
+                }
+            }
+            (TypeSignature::OptionalType(og_subty), TypeSignature::OptionalType(target_subty)) => {
+                let (variant_local, sub_locals) = locals.split_first().ok_or_else(|| {
+                    GeneratorError::InternalError(
+                        "Not enough locals for duck-typing an optional".to_owned(),
+                    )
+                })?;
+                self.duck_type_stack(builder, og_subty, target_subty, sub_locals)?;
+                builder.local_set(*variant_local);
+            }
+            (TypeSignature::ResponseType(og_subty), TypeSignature::ResponseType(target_subty)) => {
+                let (og_ok_ty, og_err_ty) = og_subty.as_ref();
+                let (target_ok_ty, target_err_ty) = target_subty.as_ref();
+
+                let (variant_local, inner_locals) = locals.split_first().ok_or_else(|| {
+                    GeneratorError::InternalError(
+                        "Not enough locals for duck-typing a response".to_owned(),
+                    )
+                })?;
+                let (ok_locals, err_locals) = inner_locals
+                    .split_at_checked(clar2wasm_ty(target_ok_ty).len())
+                    .ok_or_else(|| {
+                        GeneratorError::InternalError(
+                            "Not enough locals for duck-typing a response".to_owned(),
+                        )
+                    })?;
```

Comment:
Should we add a check to validate that `locals.len()` matches the full expected count? Something like:
```
let expected = clar2wasm_ty(target_ty).len();
if locals.len() != expected { <throw_error> };
```
This might ensure that all splits (`split_first() and split_at_checked()`) are safe and protects against type shape mismatches. What you think?

Reply:
The number of locals generated being exactly what's needed for the full value, I do not see the point in having this check.
More than that, the code is tested, so we know (except for lists for now) that the assignation to locals are correct.
Can you imagine a snippet that would provoke a bug for which this check should be mandatory?

Reply:
If the function changes, we will need a recompilation anyway, it's not gonna be dynamic. So I don't see how that would be an issue :thinking:

Reply:
Ok, your implementation is well covered with tests anyway. Let's move this PR forward. Thanks!